File: scripts/strip-html-bom.mjs
/**
 * Remove stray ? or BOM before injected header/footer.
 * Run: node scripts/strip-html-bom.mjs
 */
import { readFileSync, writeFileSync } from "fs"
import { globSync } from "fs"
import { join, dirname } from "path"
import { fileURLToPath } from "url"

const ROOT = join(dirname(fileURLToPath(import.meta.url)), "..")

const files = globSync("**/*.html", { cwd: ROOT }).filter(
  (f) => !f.startsWith("partials/") && !f.includes("faq-accordion-fragment")
)

let count = 0
for (const rel of files) {
  const path = join(ROOT, rel)
  let buf = readFileSync(path)
  if (buf[0] === 0xef && buf[1] === 0xbb && buf[2] === 0xbf) {
    buf = buf.subarray(3)
  }
  let text = buf.toString("utf8")
  const next = text
    .replace(/\ufeff    <header/g, "    <header")
    .replace(/\ufeff    <footer/g, "    <footer")
    .replace(/\?    <header/g, "    <header")
    .replace(/\?    <footer/g, "    <footer")
  if (next !== text) {
    writeFileSync(path, next, "utf8")
    count++
    console.log("fixed", rel)
  }
}
console.log(`Done. ${count} file(s).`)

File: scripts/strip-html-bom.py
#!/usr/bin/env python3
"""Remove UTF-8 BOM / stray ? prefixes before injected header/footer in HTML files."""

import glob
from pathlib import Path

ROOT = Path(__file__).resolve().parents[1]


def strip_bom_bytes(path: Path) -> bool:
    raw = path.read_bytes()
    if raw.startswith(b"\xef\xbb\xbf"):
        path.write_bytes(raw[3:])
        return True
    return False


def clean_html(path: Path) -> bool:
    text = path.read_text(encoding="utf-8-sig")
    original = text
    text = text.replace("\ufeff    <header", "    <header")
    text = text.replace("\ufeff    <footer", "    <footer")
    text = text.replace("?    <header", "    <header")
    text = text.replace("?    <footer", "    <footer")
    if text != original:
        path.write_text(text, encoding="utf-8", newline="\n")
        return True
    return False


def main():
    changed = []
    for rel in ["partials/header.html", "partials/footer.html"]:
        p = ROOT / rel
        if p.exists() and strip_bom_bytes(p):
            changed.append(f"BOM removed: {rel}")

    for f in glob.glob(str(ROOT / "**/*.html"), recursive=True):
        if "partials" in f or "faq-accordion-fragment" in f:
            continue
        p = Path(f)
        if clean_html(p):
            changed.append(p.relative_to(ROOT).as_posix())

    print(f"Cleaned {len(changed)} file(s)")
    for line in changed:
        print(" ", line)


if __name__ == "__main__":
    main()
